context/src/main/java/io/grpc/Context.java

```diff
@@ -878,11 +876,21 @@ public String toString() {
    */
   public abstract static class Storage {
     /**
+     * @deprecated This is an old API that is no longer used.
+     */
+    @Deprecated
+    public void attach(Context toAttach) {
+      throw new UnsupportedOperationException("Deprecated. Do not call.");
+    }
+
+    /**
      * Implements {@link io.grpc.Context#attach}.
      *
      * @param toAttach the context to be attached
+     * @return A {@link Context} that should be passed back into {@link #detach(Context, Context)}
+     *        as the {@code toRestore} parameter.
      */
-    public abstract void attach(Context toAttach);
+    public abstract Context doAttach(Context toAttach);
```

Comment:
Adding a new method is not binary compatible. You can provide a default implementation that uses attach().

Note that because of the current definition of current(), you'd have to convert `null` to `ROOT` within doAttach().